docs/how_to/deploy/adreno.rst

```diff
@@ -65,134 +78,442 @@ Reasons of using textures:
+can connect to this tracker and acquire a remote device handle to communicate.
 
-   target="opencl -device=adreno"
+**TVM RPC:**
+TVM RPC is a native application that runs on the remote device (Android in our case) and registers itself to the TVM Tracker
+running on the host.
 
-Let's write a simple model with one convolutional (conv2d) layer and take a look at generated kernels for these
-two targets
 
-.. code:: python
+Hence, for RPC based setup we will have above components running on host and target device. Below sections explain how to setup the same
+manually and also inside docker using automated tools.
 
-   import tvm
-   from tvm import relay
-   import numpy as np
+**Automated RPC Setup:**
+Here we will explain how to setup RPC in docker environment.
 
-   input_shape=(1, 56, 56, 32)
-   filter_shape=(3, 3, 32, 64)
-   filter = np.random.rand(*filter_shape)
+Below command launches tracker in docker environment, where tracker listens on port 9190.
 
-   dtype="float32"
-   input = tvm.relay.var("input", shape=input_shape, dtype=dtype)
-   weight = tvm.relay.var("weight", shape=filter_shape, dtype=dtype)
-   D = relay.nn.conv2d(input, weight, padding=(1, 1), data_layout="NHWC", kernel_layout="HWIO", out_dtype=dtype)
+::
 
-   mod = relay.Function([input, weight], D)
-   params = {
-      "weight": tvm.nd.array(filter)
-   }
+   ./tests/scripts/ci.py adreno -i # Launch a new shell on the anreno docker
+   source  tests/scripts/setup-adreno-env.sh -e tracker -p 9190
 
-Now compile our model with the classic OpenCL target and print its modules:
+Now, the below comand can run TVM RPC on remote android device with id "abcdefgh".
 
-.. code:: python
 
-   target="opencl"
+::
 
-   with tvm.transform.PassContext(opt_level=3):
-      graph, lib, params = relay.build_module.build(mod, target, params=params)
-   print(lib.imported_modules[0].get_source())
+   ./tests/scripts/ci.py adreno -i # Launch a new shell on adreno docker.
+   source  tests/scripts/setup-adreno-env.sh -e device -p 9190 -d abcdefgh
 
-Notice that the generated convolution kernel has pointers in
-the initialization of the function. The kernels generated with the above target are buffer-based.
 
-.. code:: c
+**Manual RPC Setup:**
 
-   __kernel void tvmgen_default_fused_nn_conv2d_kernel0(__global float* restrict p0, __global double* restrict p1, __global float* restrict conv2d_nhwc) {
-   // body..
+Please refer to the tutorial
+`How To Deploy model on Adreno <https://tvm.apache.org/docs/how_to/deploy_models/deploy_model_on_adreno.html>`_
+for manual RPC environment setup.
+
+This concludes RPC Setup and we have rpc-tracker available on host 127.0.0.1 (rpc-tracker) and port 9190 (rpc-port).
+
+
+.. _commandline_interface:
+
+Commandline Tools
+-----------------
+
+Here we describe entire compilation process using command line tools. TVM has command line utility "tvmc" to perform
```

Comment:
I completely agree that using of generic tool like tvmc is better than development of python script, on the other hand I have not seen that tvmc can convert model to fp16. While it is the must have feature for running of model on GPU/Adreno.

If tvmc cannot transform model to fp16, it's better not to promote it here

Reply:
fp16 is performance related aspect of Adreno and I think we should promote it.

```tvmc``` at the moment supports post processing (after import) only for ```BYOC``` target options. I was thinking of enhancing ```tvmc``` to allow vendor specific post processing before ```relay.build```.  

I think we can leave it this for now and rectify the same after ```tvmc``` changes.

What do you think ?

Reply:
I would not promote tvmc in adreno flow until it support conversion to fp16.

If you are planning to add such kind of support, let's postpone these changes until tvmc get this support.

I.e. I propose to modify "Deploy" section without mentioning of tvmc.

In other case it will be formally correct, such steps are easy and clear but useless. No one will deploy fp32 on GPU.

Reply:
I think ```tvmc``` command line way helps new developers a lot. hence, I worked on enhancing tvmc to support pre and post processing hooks https://github.com/apache/tvm/pull/14010